lib/Feature/FeatureModelParser.cpp

```diff
@@ -225,4 +252,361 @@ std::unique_ptr<xmlDoc, void (*)(xmlDocPtr)> FeatureModelXmlParser::parseDoc() {
+    for (auto & OrGroup : OrGroupMapping) {
+      FMB.addRelationship(std::get<1>(OrGroup.second), std::get<2>(OrGroup.second), std::get<0>(OrGroup.second));
+    }
+  }
+
+  return true;
+}
+
+/// This method extracts the cardinality from the given line.
+/// The cardinality is wrapped in square brackets (e.g., [1,1])
+/// \param StringToExtractFrom the string to extract the cardinality from
+/// \return
+std::optional<std::tuple<int, int>> FeatureModelSxfmParser::extractCardinality(const string& StringToExtractFrom) {
+  std::optional<int> MinCardinality;
+  std::optional<int> MaxCardinality;
+
+  // Search for the first occurrence of '['; then read in the min cardinality
+  // until the comma. Afterwards, read in the max cardinality until ']'
+  std::string::size_type Pos = StringToExtractFrom.find_first_of('[');
+  if (Pos == std::string::npos) {
+    std::cerr << "No cardinality given in or group!" << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+  MinCardinality = parseCardinality(readUntil(StringToExtractFrom, ',', Pos + 1));
+  Pos = StringToExtractFrom.find_first_of(',', Pos);
+  MaxCardinality = parseCardinality(readUntil(StringToExtractFrom, ']', Pos + 1));
+
+  if (!MinCardinality.has_value() || !MaxCardinality.has_value()) {
+    std::cerr << "No parsable cardinality!" << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+
+  if (MinCardinality.value() != 1 || (MaxCardinality.value() != 1 && MaxCardinality.value() != UINT_MAX)) {
+    std::cerr << "Cardinality unsupported. We support cardinalities [1,1] (alternative) or [1, *] (or group)." << std::endl;
+    return std::optional<std::tuple<int, int>>();
+  }
+
+  return std::optional<std::tuple<int, int>>(std::tuple<int, int>{MinCardinality.value(), MaxCardinality.value()});
+}
+
+
+/// This method parses the given cardinality and returns an optional.
+/// If the optional is empty, the process failed; otherwise the result contains
+/// either UINT_MAX for the wildcard or the cardinality number as integer.
+/// \param CardinalityString the cardinality to parse
+/// \return an optional that contains no integer in case of failure or UINT_MAX
+/// for wildcard, or the number itself.
+std::optional<int> FeatureModelSxfmParser::parseCardinality(const string& CardinalityString) {
+  std::optional<int> Result = std::optional<int>();
+  if (CardinalityString == "*") {
+    // We use UINT_MAX as our magic integer (which is -1 as int) to indicate
+    // that the cardinality is a wildcard.
+    Result = UINT_MAX;
+  } else {
+    // Convert the string into an integer in a safe way
+    char *End;
+    long LongNumber;
+    errno = 0;
+    LongNumber = strtol(CardinalityString.c_str(), &End, 0);
+    if (errno == ERANGE || LongNumber < INT_MIN || LongNumber > INT_MAX
+        || *CardinalityString.c_str() == '\0' || *End != '\0') {
+      std::cerr << "The following cardinality is not integer: " << CardinalityString << std::endl;
```

Comment:
```suggestion
      llvm::errs() << "The following cardinality is not integer: " << CardinalityString << '\n';
```
We try to use the llvm out and error stream because they are a bit easier to use and have some interesting features. So, to be consistent, it would be good to use them everywhere.